File: Java/src/World/body/animal/Fox.java
package World.body.animal;

import GUI.Color_obj;
import World.Point;
import World.World;
import World.body.Animal;
import World.body.Body;

import java.io.Serializable;

public class Fox extends Animal implements Serializable {
    private static final long serialVersionUID = 123456789L;
    public Fox(Point point_location, World world) {
        super(point_location, world);
        setName("Fox");
        setInitiative(7);
        setPower(3);
        setColor(Color_obj.FOX);
    }

    @Override
    protected boolean repel_attack(Body attacker) {
        return false;
    }

    @Override
    protected void action() {
        Point point = new Point(getPoint_location().getX(), getPoint_location().getY());
        Point new_point = new Point(getPoint_location().getX(), getPoint_location().getY());
        boolean tmp = true;
        int attempt=0;
        while(tmp){
            random_location(new_point);
            tmp=false;
            Body body = getWorld().get_body(new_point);
            if(body instanceof Animal){
                if(body.getPower()>getPower()){
                    new_point.setX(point.getX());
                    new_point.setY(point.getY());
                    tmp = true;
                    attempt++;
                    if(attempt>12){
                        tmp = false;
                    }
                }
            }
        }
        setLast_position(point);
        setPoint_location(new_point);
    }
}

File: Python/World/Body/Animal/fox.py
from GUI.colors import Colors
from World.Body.animal import Animal
from World.point import Point


class Fox(Animal):
    def __init__(self, point_location, world):
        super().__init__(point_location, world)
        self.set_name("Fox")
        self.set_initiative(7)
        self.set_power(3)
        self.set_color(Colors.FOX)

    def repel_attack(self, attacker):
        return False

    def action(self):
        point = Point(self.get_point_location().x, self.get_point_location().y)
        new_point = Point(self.get_point_location().x, self.get_point_location().y)
        tmp = True
        attempt = 0
        while tmp:
            self.random_location(new_point)
            tmp = False
            body = self.world.get_body(new_point)
            if isinstance(body, Animal):
                if body.get_power() > self.get_power():
                    new_point.x = point.x
                    new_point.y = point.y
                    tmp = True
                    attempt += 1
                    if attempt > 12:
                        tmp = False
        self.set_last_position(point)
        self.set_point_location(new_point)